File: client/app/src/main/java/me/shouheng/notepal/fragment/base/BaseModelFragment.java
package me.shouheng.notepal.fragment.base;

import android.arch.lifecycle.ViewModelProviders;
import android.databinding.ViewDataBinding;
import android.support.annotation.ColorInt;
import android.text.TextUtils;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import org.polaric.colorful.PermissionUtils;

import java.util.LinkedList;
import java.util.List;

import me.shouheng.notepal.R;
import me.shouheng.notepal.activity.base.CommonActivity;
import me.shouheng.notepal.dialog.CategoryEditDialog;
import me.shouheng.notepal.dialog.picker.CategoryPickerDialog;
import me.shouheng.notepal.manager.LocationManager;
import me.shouheng.notepal.model.Category;
import me.shouheng.notepal.model.Location;
import me.shouheng.notepal.model.Model;
import me.shouheng.notepal.model.ModelFactory;
import me.shouheng.notepal.provider.CategoryStore;
import me.shouheng.notepal.util.NetworkUtils;
import me.shouheng.notepal.util.ToastUtils;
import me.shouheng.notepal.util.ViewUtils;
import me.shouheng.notepal.viewmodel.CategoryViewModel;
import me.shouheng.notepal.widget.FlowLayout;

/**
 * Created by wangshouheng on 2017/9/3.
 */
public abstract class BaseModelFragment<T extends Model, V extends ViewDataBinding> extends BaseFragment<V> {

    // region Base logic about category

    private List<Category> allCategories;
    private List<Category> cs = new LinkedList<>();

    private CategoryEditDialog categoryEditDialog;

    /**
     * Call this method and override {@link #onGetSelectedCategories(List)} to implement
     * the logic of getting categories.
     *
     * @param selected selected categories
     */
    protected void showCategoriesPicker(List<Category> selected) {
        List<Category> all = getAllCategories();

        // try to avoid NPE
        if (selected == null) selected = new LinkedList<>();
        cs = selected;

        for (Category c : selected) {
            for (Category a : all) {
                if (c.getCode() == a.getCode()) {
                    a.setSelected(true);
                }
            }
        }

        CategoryPickerDialog dialog = CategoryPickerDialog.newInstance(all);
        dialog.setOnConfirmClickListener(this::onGetSelectedCategories);
        dialog.setOnAddClickListener(this::showCategoryEditor);
        dialog.show(getFragmentManager(), "CATEGORY_PICKER");
    }

    private void showCategoryEditor() {
        categoryEditDialog = CategoryEditDialog.newInstance(ModelFactory.getCategory(), this::saveCategory);
        categoryEditDialog.show(getFragmentManager(), "CATEGORY_EDIT_DIALOG");
    }

    private void saveCategory(Category category) {
        ViewModelProviders.of(this).get(CategoryViewModel.class)
                .saveModel(category).observe(this, categoryResource -> {
            if (categoryResource == null) {
                ToastUtils.makeToast(R.string.text_error_when_save);
                return;
            }
            switch (categoryResource.status) {
                case SUCCESS:
                    ToastUtils.makeToast(R.string.text_save_successfully);
                    allCategories.add(0, category);
                    showCategoriesPicker(cs);
                    break;
                case FAILED:
                    ToastUtils.makeToast(R.string.text_error_when_save);
                    break;
            }
        });
    }

    protected void onGetSelectedCategories(List<Category> categories) {
    }

    private List<Category> getAllCategories() {
        if (allCategories == null) {
            allCategories = CategoryStore.getInstance(getContext()).get(null, null);
        }
        return allCategories;
    }

    /**
     * Call this method and override {@link #getTagsLayout()} to implement the logic of showing tags.
     *
     * @param stringTags tags string
     */
    protected void addTagsToLayout(String stringTags) {
        if (getTagsLayout() == null) return;
        getTagsLayout().removeAllViews();
        if (TextUtils.isEmpty(stringTags)) return;
        String[] tags = stringTags.split(CategoryViewModel.CATEGORY_SPLIT);
        for (String tag : tags) addTagToLayout(tag);
    }

    protected FlowLayout getTagsLayout() {
        return null;
    }

    private void addTagToLayout(String tag) {
        if (getTagsLayout() == null) return;

        int margin = ViewUtils.dp2Px(getContext(), 2f);
        int padding = ViewUtils.dp2Px(getContext(), 5f);
        TextView tvLabel = new TextView(getContext());
        tvLabel.setLayoutParams(new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.WRAP_CONTENT));
        ViewGroup.MarginLayoutParams params = new ViewGroup.MarginLayoutParams(tvLabel.getLayoutParams());
        params.setMargins(margin, margin, margin, margin);
        tvLabel.setLayoutParams(params);
        tvLabel.setPadding(padding, 0, padding, 0);
        tvLabel.setBackgroundResource(R.drawable.label_background);
        tvLabel.setText(tag);

        getTagsLayout().addView(tvLabel);
    }

    // endregion
    // region location

    protected void tryToLocate() {
        if (!NetworkUtils.isNetworkAvailable(getActivity())) {
            ToastUtils.makeToast(R.string.check_network_availability);
            return;
        }
        if (getActivity() != null) {
            PermissionUtils.checkLocationPermission((CommonActivity) getActivity(), this::baiduLocate);
        }
    }

    private void baiduLocate() {
        ToastUtils.makeToast(R.string.trying_to_get_location);
        LocationManager.getInstance().locate();
    }

    protected void onGetLocation(Location location) {
    }

    // endregion

    public void onColorSelection(@ColorInt int i) {
        if (categoryEditDialog != null) {
            categoryEditDialog.updateUIBySelectedColor(i);
        }
    }
}

File: client/app/src/main/java/me/urakalee/next2/base/fragment/BaseModelFragment.kt
package me.urakalee.next2.base.fragment

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.support.annotation.ColorInt
import android.text.TextUtils
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import me.shouheng.notepal.R
import me.shouheng.notepal.dialog.CategoryEditDialog
import me.shouheng.notepal.dialog.picker.CategoryPickerDialog
import me.shouheng.notepal.model.Category
import me.shouheng.notepal.model.Model
import me.shouheng.notepal.model.ModelFactory
import me.shouheng.notepal.model.data.LoadStatus
import me.shouheng.notepal.provider.CategoryStore
import me.shouheng.notepal.util.ToastUtils
import me.shouheng.notepal.util.ViewUtils
import me.shouheng.notepal.viewmodel.CategoryViewModel
import me.shouheng.notepal.widget.FlowLayout

/**
 * @author Uraka.Lee
 */
abstract class BaseModelFragment<T : Model> : BaseFragment() {

    // region Base logic about category

    private var allCategories: MutableList<Category>? = null
    private var categories: List<Category> = listOf()

    private var categoryEditDialog: CategoryEditDialog? = null

    protected open val tagsLayout: FlowLayout?
        get() = null

    /**
     * Call this method and override [.onGetSelectedCategories] to implement
     * the logic of getting categories.
     *
     * @param selected selected categories
     */
    protected fun showCategoriesPicker(selected: List<Category>?) {
        val allCategories = getAllCategories()
        val selectedCategories = selected ?: listOf()
        categories = selectedCategories

        for (category in selectedCategories) {
            for (item in allCategories) {
                if (category.code == item.code) {
                    item.isSelected = true
                }
            }
        }

        val dialog = CategoryPickerDialog.newInstance(allCategories)
        dialog.setOnConfirmClickListener { this.onGetSelectedCategories(it) }
        dialog.setOnAddClickListener { this.showCategoryEditor() }
        dialog.show(fragmentManager, "CATEGORY_PICKER")
    }

    private fun getAllCategories(): List<Category> {
        if (allCategories == null) {
            allCategories = CategoryStore.getInstance(context).get(null, null)
        }
        return allCategories ?: listOf()
    }

    protected open fun onGetSelectedCategories(categories: List<Category>) {}

    private fun showCategoryEditor() {
        categoryEditDialog = CategoryEditDialog.newInstance(ModelFactory.getCategory()) { this.saveCategory(it) }
        categoryEditDialog?.show(fragmentManager, "CATEGORY_EDIT_DIALOG")
    }

    private fun saveCategory(category: Category) {
        ViewModelProviders.of(this).get(CategoryViewModel::class.java)
                .saveModel(category)
                .observe(this, Observer { categoryResource ->
                    if (categoryResource == null) {
                        ToastUtils.makeToast(R.string.text_error_when_save)
                        return@Observer
                    }
                    when (categoryResource.status) {
                        LoadStatus.SUCCESS -> {
                            ToastUtils.makeToast(R.string.text_save_successfully)
                            allCategories?.add(0, category)
                            showCategoriesPicker(categories)
                        }
                        LoadStatus.FAILED -> {
                            ToastUtils.makeToast(R.string.text_error_when_save)
                        }
                        else -> {
                            // pass
                        }
                    }
                })
    }

    /**
     * Call this method and override [.getTagsLayout] to implement the logic of showing tags.
     *
     * @param stringTags tags string
     */
    protected fun addTagsToLayout(stringTags: String) {
        if (tagsLayout == null) return
        tagsLayout!!.removeAllViews()
        if (TextUtils.isEmpty(stringTags)) return
        val tags = stringTags.split(CategoryViewModel.CATEGORY_SPLIT.toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        for (tag in tags) addTagToLayout(tag)
    }

    private fun addTagToLayout(tag: String) {
        if (tagsLayout == null) return

        val margin = ViewUtils.dp2Px(context!!, 2f)
        val padding = ViewUtils.dp2Px(context!!, 5f)
        val label = TextView(context)
        label.layoutParams = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.WRAP_CONTENT)
        val params = ViewGroup.MarginLayoutParams(label.layoutParams as LinearLayout.LayoutParams)
        params.setMargins(margin, margin, margin, margin)
        label.layoutParams = params
        label.setPadding(padding, 0, padding, 0)
        label.setBackgroundResource(R.drawable.label_background)
        label.text = tag

        tagsLayout!!.addView(label)
    }

    // endregion

    fun onColorSelection(@ColorInt i: Int) {
        if (categoryEditDialog != null) {
            categoryEditDialog!!.updateUIBySelectedColor(i)
        }
    }
}
